apollo-federation/src/sources/connect/validation/graphql/strings.rs

```diff
@@ -0,0 +1,158 @@
+//! Helpers for dealing with GraphQL literal strings and getting locations within them.
+//!
+//! Specifically, working around these issues with `apollo-compiler` which make determining the
+//! line/column of locations _within_ a string impossible:
+//!
+//! 1. Quotes are included in locations (i.e., `line_column_range`) but not the values from `Node<Value>`.
+//! 2. String values in `Node<Value>` are stripped of insignificant whitespace
+
+use std::ops::Range;
+
+use apollo_compiler::ast::Value;
+use apollo_compiler::parser::LineColumn;
+use apollo_compiler::parser::SourceMap;
+use apollo_compiler::Node;
+use line_col::LineColLookup;
+
+#[derive(Clone, Copy)]
+pub(crate) struct GraphQLString<'schema> {
+    /// The GraphQL String literal without quotes, but with all whitespace intact
+    raw_string: &'schema str,
+    /// Where `raw_string` _starts_ in the source text
+    offset: usize,
+}
+
+impl<'schema> GraphQLString<'schema> {
+    /// Get the raw string value of this GraphQL string literal
+    ///
+    /// Returns `None` if the value was not a string literal or if location data is messed up
+    pub fn new(value: &'schema Node<Value>, sources: &'schema SourceMap) -> Option<Self> {
```

Comment:
Nitpick: idiomatically, this seems more like a `Result` than an `Option`, since the "messed up" cases would be an error. All callers also use `ok_or...` to convert to a `Result`.